bids/visualize.py

```diff
@@ -0,0 +1,758 @@
+import pandas as pd
+import plotly.express as px
+import plotly.graph_objects as go
+from plotly.subplots import make_subplots
+
+
+class EventPlotter:
+    """Class to handle plotting the content of events files.
+
+    Parameters
+    ----------
+    events_file : str | Path
+        Path to the events file.
+
+    event_column : str | None, optional
+        Name of the column containing the events.
+        If `None`, events types are assumed to be in the "trial_type" column.
+
+    include : list[str] | None, optional
+        List of events type to include in the plot.
+    """
+
+    def __init__(
+        self,
+        events_file: str | Path,
+        event_column: str | None = None,
+        include: list[str] | None = None,
+    ):
+        self.COLOR_LIST = px.colors.qualitative.Plotly
+        self.FONT_SIZE = 14
+        self.STANDOFF = 16
+        self.LINE_WIDTH = 3
+        self.GRID_COLOR = "black"
+        self.AXES_COLOR = "black"
+        self.AXES_LINE_WIDTH = 2
+        self.BG_COLOR = "rgb(255,255,255)"
+        self.EVENT_HEIGHT = 1
+        self.FAST_RESPONSE_THRESHOLD = 0.2
+        self.NB_BINS = 40
+        self.TWO_COLUMN_WIDTHS = [0.7, 0.3]
+        self.THREE_COLUMN_WIDTHS = [0.7, 0.15, 0.15]
+        self.FOUR_COLUMN_WIDTHS = [0.7, 0.1, 0.1, 0.1]
+        self.TICK_LENGTH = 6
+
+        self.title: None | str = None
+        self.fig = None
+
+        self._trial_type_index: int = 0
+        self._trial_types: None | list[str] = None
+        self._bottom_row: dict(str, list[int]) = {
+            "ISI": [],
+            "duration": [],
+            "response": [],
+        }
+
+        self._get_data_from_file(events_file)
+
+        self.event_column = event_column
+        if event_column is None:
+            self.event_column = "trial_type"
+
+        if len(self.data.columns.to_list()) == 2:
+            warnings.warn(
+                f"""Only columns 'onset' and 'duration' found in:
+    {events_file}
+Creating a dummy trial_type column.
+            """
+            )
+            self.data["trial_type"] = "trial_type"
+
+        self.trial_types(include=include)
+        if self.trial_types() is None or self.nb_trial_types == 0:
+            warnings.warn(
+                f"No trial types found in {events_file} for 'include={include}'"
+            )
+            return
+
+        if self.nb_trial_types > 13:
+            warnings.warn(
+                f"""More than 13 trial types found in {events_file}.
+The plot will be unreadable.
+
+Specify a subset of trial types using the 'include' argument.
+                """
+            )
+            return
+
+        self.fig = go.FigureWidget(
+            make_subplots(
+                rows=len(self.trial_types()),
+                cols=self.nb_cols,
+                horizontal_spacing=0.02,
+                vertical_spacing=0.08,
+                shared_xaxes=True,
+                column_widths=self.column_widths,
+            )
+        )
```

Comment:
Is this expensive? You may not want it in `__init__`. You could do something like `self.fig_ = None` here and then

```Python
@property
def fig(self):
    if self.fig_ is None:
        self.fig_ = go.FigureWidget(...)
    return self.fig_
```

This way it's generated when you need it but not before. The post-underscore convention is something @adelavega introduced me to from scikit-learn, to indicate the instance variable is generated after init. On the other hand, we could also do a "hidden" variable `_fig`, if we're exposing it through the `fig` property.